### PR TITLE
Improve unit & rank search filters

### LINKED ISSUE #59: Make "Currently on the force" a field available when browsing
This builds off of #55, but might require what @sarayourfriend suggested: adding this as a property on officers. I haven't looked at how searching works though so I'm not certain on the best approach.

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Description of Changes
This PR attempts to improve the unit and rank search filters by unifying their behavior and making them scrollable. OpenOversight was probably never intended for as many ranks/units we have (over 100). Currently the ranks are a multi-select checkbox list, while the units are a drop down list. This is quite cumbersome for users and prevents them from being able to select more than one unit.

I've modified the unit field so that it is also a multi-select checkbox list. When both of these search fields are opened the page becomes tremendously long rendering all of the results. In order to prevent this, I changed the frontend to allow scrolling overflow for these 2 options.

I also altered which fields show up in the officer list entry to better match which fields users can (and care about) searching against.

Lastly, I also moved the `currently_on_force` template filter added in #55 to a function on the `Officer` model (now that I had figured out where to add it!). This resolves #59.

This work is based off the request from #56, though I don't feel like a text search here would be appropriate. We want users to be able to select multiple ranks/units, but we also don't want to leave them guessing what the ranks & units are. This provides a suitable middle ground in my opinion: they don't have to know what the ranks/units a priori, and they can select multiple ranks/units when making a search.

#### List entry
Screenshot_2021-12-29_08-46-29

#### Dropdowns

https://user-images.githubusercontent.com/10214785/147685251-0845f84b-d133-4efd-9a8e-382fb1edb147.mp4

#### List entry
Screenshot_2021-12-29_08-47-30

#### Dropdowns

https://user-images.githubusercontent.com/10214785/147685271-4067fd07-d80d-4d07-9573-975a788158eb.mp4

## Tests and linting

 - [x] I have rebased my changes on `main`

 - [x] `just lint` passes

 - [x] `just test` passes

### CODE REVIEW

File: OpenOversight/app/main/views.py [hunk 1]
```diff
@@ -608,9 +608,9 @@ def edit_department(department_id):
 def list_officer(
     department_id,
     page=1,
-    race=[],
-    gender=[],
-    rank=[],
+    race=None,
+    gender=None,
+    rank=None,
```
Comment (sarayourfriend) [inline OpenOversight/app/main/views.py:613]: Why this change?

File: OpenOversight/app/models.py [hunk 2]
```diff
@@ -188,12 +188,25 @@ def job_title(self):
...
             return max(
                 self.assignments_lazy, key=lambda x: x.star_date or date.min
             ).star_no
 
+    def currently_on_force(self):
+        if self.assignments_lazy:
+            most_recent = max(
+                self.assignments_lazy, key=lambda x: x.star_date or date.min
+            )
+            return "Yes" if most_recent.resign_date is None else "No"
```
Comment (sarayourfriend) [inline OpenOversight/app/models.py:208]: This is missing the "Uncertain" branch from the original implementation. Not sure if it matters?

```suggestion
        if self.assignments_lazy:
            most_recent = max(
                self.assignments_lazy, key=lambda x: x.star_date or date.min
            )
            return "Yes" if most_recent.resign_date is None else "No"
        return "Uncertain"
```

Oh never mind, I see it's in the template now... 😕 I think I prefer that kind of thing in the property function itself cause it makes it actually unit testable, but I suppose it's better to just follow the already established pattern.

File: OpenOversight/app/templates/list_officer.html [hunk 3]
```diff
@@ -98,13 +98,15 @@ <h3 class="panel-title accordion-toggle">Rank</h3>
             <h3 class="panel-title accordion-toggle">Unit</h3>
           </div>
           <div class="collapse" id="filter-unit">
-            <div class="panel-body">
+            <div class="panel-body panel-body-long">
               <div class="form-group">
-                <select class="form-control" name="unit">
```
Comment (sarayourfriend) [inline OpenOversight/app/templates/list_officer.html:103]: `select` actually has a `multiple` attribute that might be perfect for this instead of a checkbox and extra CSS: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#attr-multiple

IIRC it will render it as scrollable text box of a fixed height. Though might not be as intuitive to some folks how to select multiple (you have to control + click to select more than one... though most folks with moderate computer literacy will have learned that trick I think? Not sure)

File: OpenOversight/app/utils.py [hunk 4]
```diff
@@ -348,20 +348,29 @@ def filter_by_form(form_data, officer_query, department_id=None):
...
+    if form_data.get("unit"):
+        unit_ids = [
+            unit.id
+            for unit in Unit.query.filter_by(department_id=department_id)
+            .filter(Unit.descrip.in_(form_data.get("unit")))
+            .all()
+        ]
+
+        if "Not Sure" in form_data["unit"]:
+            form_data["unit"].append(None)
```
Comment (sarayourfriend) [inline OpenOversight/app/utils.py:364]: How does this work exactly? This might warrant a comment to explain why appending `None` is necessary.

File: OpenOversight/app/main/views.py (hunk 1 above)
Comment (AetherUnbound) [reply OpenOversight/app/main/views.py:613]: This falls into the ["mutable default arguments" problem](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments).

File: OpenOversight/app/models.py (hunk 2 above)
Comment (AetherUnbound) [reply OpenOversight/app/models.py:208]: Ah, that's a great point! I'll add that

File: OpenOversight/app/utils.py (hunk 4 above)
Comment (AetherUnbound) [reply OpenOversight/app/utils.py:364]: Great call, can do!

File: OpenOversight/app/utils.py [hunk 5]
```diff
@@ -361,6 +361,7 @@ def filter_by_form(form_data, officer_query, department_id=None):
         ]
 
         if "Not Sure" in form_data["unit"]:
+            # Convert "Not Sure" to None, so the resulting SQL query allows NULL values
```
Comment (sarayourfriend) [inline OpenOversight/app/utils.py:364]: This is kind of confusing to me... should it _replace_ `"Not Sure"`? Cause at this point we're just adding `None` to allow null values... and I suppose `Not Sure` will never match anything, so the `in` query still works fine, right? So this is good 😂

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
This crashes locally with:

```
web_1           | Traceback (most recent call last):
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2088, in __call__
web_1           |     return self.wsgi_app(environ, start_response)
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
web_1           |     response = self.handle_exception(e)
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2070, in wsgi_app
web_1           |     response = self.full_dispatch_request()
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1515, in full_dispatch_request
web_1           |     rv = self.handle_user_exception(e)
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1513, in full_dispatch_request
web_1           |     rv = self.dispatch_request()
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1499, in dispatch_request
web_1           |     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
web_1           |   File "/usr/src/app/OpenOversight/app/main/views.py", line 743, in list_officer
web_1           |     return render_template(
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/templating.py", line 147, in render_template
web_1           |     return _render(
web_1           |   File "/usr/local/lib/python3.9/site-packages/flask/templating.py", line 128, in _render
web_1           |     rv = template.render(context)
web_1           |   File "/usr/local/lib/python3.9/site-packages/jinja2/environment.py", line 1291, in render
web_1           |     self.environment.handle_exception()
web_1           |   File "/usr/local/lib/python3.9/site-packages/jinja2/environment.py", line 925, in handle_exception
web_1           |     raise rewrite_traceback_stack(source=source)
web_1           |   File "/usr/src/app/OpenOversight/app/templates/list_officer.html", line 2, in top-level template code
web_1           |     {% import "bootstrap/wtf.html" as wtf %}
web_1           |   File "/usr/src/app/OpenOversight/app/templates/base.html", line 104, in top-level template code
web_1           |     {% block content %}{% endblock %}
web_1           |   File "/usr/src/app/OpenOversight/app/templates/list_officer.html", line 160, in block 'content'
web_1           |     <dd>{{ officer.unit_descrip()|default('Unknown') }}</dd>
web_1           |   File "/usr/src/app/OpenOversight/app/models.py", line 193, in unit_descrip
web_1           |     return max(
web_1           | AttributeError: 'NoneType' object has no attribute 'descrip'
```

I guess this happens when there is no unit associated with the assignment?

For context I've just run `just fresh-start && just up`.

File: OpenOversight/app/utils.py (hunk 5 above)
Comment (AetherUnbound) [reply OpenOversight/app/utils.py:364]: I think so? :sweat_smile: I'm no so familiar with this code...but yea I think it works as you describe!

**Comment by AetherUnbound** [comment]
Gah, fixed!
